### PR TITLE
📦 NEW: Add eg-policy hardtanh ctx architecture

### PR DESCRIPTION
**Pull request by ianagbip1oti**
```
-------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 3.80 +/- 3.02, nElo: 7.27 +/- 5.78
LOS: 99.32 %, DrawRatio: 52.45 %, PairsRatio: 1.08
Games: 13898, Wins: 2922, Losses: 2770, Draws: 8206, Points: 7025.0 (50.55 %)
Ptnml(0-2): [66, 1526, 3645, 1614, 98], WL/DD Ratio: 0.44
LLR: 2.91 (100.7%) (-2.25, 2.89) [0.00, 3.00]
--------------------------------------------------

-------------------------------------------------
Results of princhess vs princhess-main (40+0.4, 1t, 128MB, DFRC_4852_v1.epd):
Elo: 6.40 +/- 5.58, nElo: 11.56 +/- 10.08
LOS: 98.77 %, DrawRatio: 48.73 %, PairsRatio: 1.16
Games: 4560, Wins: 1029, Losses: 945, Draws: 2586, Points: 2322.0 (50.92 %)
Ptnml(0-2): [39, 503, 1111, 589, 38], WL/DD Ratio: 0.49
LLR: 2.90 (100.4%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------

-------------------------------------------------
Results of princhess vs princhess-main (8+0.08, 1t, 128MB, UHO_Lichess_4852_v1.epd):
Elo: 0.09 +/- 2.52, nElo: 0.17 +/- 4.53
LOS: 52.87 %, DrawRatio: 50.41 %, PairsRatio: 1.01
Games: 22550, Wins: 5126, Losses: 5120, Draws: 12304, Points: 11278.0 (50.01 %)
Ptnml(0-2): [232, 2556, 5684, 2580, 223], WL/DD Ratio: 0.59
LLR: 2.89 (100.1%) (-2.25, 2.89) [-3.00, 0.00]
--------------------------------------------------
```

## Summary by CodeRabbit

* **New Features**
  * Added separate middlegame and endgame policy networks for more specialized move evaluation.
  * Engine fingerprints now report the embedded network checksums through UCI and benchmarking tools.
  * Added support for Dirichlet-based randomization and improved MCTS configuration access.
  * Expanded training tools with endgame policy training, metadata manifests, progress reporting, and learning-rate analysis.
  * Added a convenient 25k-node SPRT gain-test command.

* **Improvements**
  * Enhanced policy-network serialization, quantization, and training-data handling.
  * Added richer training metrics and network fingerprint visibility.

### CODE REVIEW

File: crates/princhess-train/src/bin/data-gen.rs [hunk 1]
```diff
@@ -1607,4 +1708,15 @@ fn main() {
...
+    let files: Vec<String> = (0..threads)
+        .map(|t| format!("data/princhess-{timestamp}-{t}.data"))
+        .collect();
+    write_toml(
+        &format!("data/princhess-{timestamp}.toml"),
+        &stats,
+        &files,
+        threads,
+        max_positions,
+    );
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/data-gen.rs:1766]: Empty files are intentional — they record that a thread slot was allocated but produced no data, which is useful for diagnosing utilisation issues. The shuffle step handles zero-byte files gracefully downstream.

File: crates/princhess-train/src/bin/data-shuffle.rs [hunk 2]
```diff
@@ -530,6 +534,121 @@ fn render_interleave_box(frame: &mut Frame, area: ratatui::layout::Rect, progres
...
+                        })
+                }
+                _ => false,
+            };
+            if all_gone {
+                fs::remove_file(&path).ok();
+            }
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/data-shuffle.rs:670]: The rescan is intentional — by re-reading the directory rather than limiting to paths from find_source_tomls, cleanup_orphaned_tomls catches orphaned TOMLs from any previous run in the same directory, not just the
  ones considered in the current shuffle pass. Restricting it to the already-discovered paths would leave behind stale TOMLs from older runs whose data files have since been deleted.

File: crates/princhess-train/src/bin/train-mg-policy.rs [hunk 3]
```diff
@@ -1074,14 +1042,13 @@ fn gradients_batch(
...
+    gradients: &mut MgPolicyNetwork,
     metrics: &mut BatchMetrics,
-    phase: Phase,
 ) {
     let state = State::from(position);
 
-    if !phase.matches(&state) {
+    if !is_training_position(&state) {
         return;
     }
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-mg-policy.rs:1048]: This is deliberate. It happens to be what gains elo and any change would require future elo checks and would be done in a separate PR.

File: src/evaluation.rs [hunk 4]
```diff
@@ -137,8 +139,11 @@ fn run_policy_net(state: &State, moves: &MoveList, t: f32) -> Vec<f32> {
         return evalns;
     }
 
-    let network = [&MG_POLICY_NETWORK, &EG_POLICY_NETWORK][usize::from(state.is_endgame())];
-    network.get_all(state, state.moves_to_indexes(moves), &mut evalns);
+    if state.is_endgame() {
+        EG_POLICY_NETWORK.get_all(state, state.moves_to_indexes(moves), &mut evalns);
+    } else {
+        MG_POLICY_NETWORK.get_all(state, state.moves_to_indexes(moves), &mut evalns);
+    }
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply src/evaluation.rs:150]: Experiments so far have shown it doesn't make the eg data worse, but in fact makes it better. This finding may not hold with current arch and data, but that requires a separate PR with sprt testing

File: src/nets.rs [hunk 5]
```diff
@@ -37,6 +52,65 @@ impl<const H: usize> Accumulator<i16, H> {
...
+    #[must_use]
+    pub fn linear_mult_relu<const Q: i32>(&self, rhs: &Accumulator<i16, H>) -> Accumulator<i16, H> {
+        let mut result = Accumulator { vals: [0i16; H] };
+
+        for i in 0..H {
+            result.vals[i] = ((i32::from(self.vals[i]) * relu(rhs.vals[i])) / Q) as i16;
+        }
+
+        result
+    }
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply src/nets.rs]: This method has been deleted

File: src/quantized_eg_policy.rs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+pub type RawCtxBias = Align16<[i16; CTX_SIZE]>;
+
+pub type RawSquareWeights = [RawLinearWeights; Square::COUNT];
+pub type RawSquareBias = [RawLinearBias; Square::COUNT];
+
+use crate::chess::Square;
+pub type QuantizedSquareSubnets =
+    QuantizedLinearNetwork<{ Square::COUNT }, INPUT_SIZE, ATTENTION_SIZE>;
+
+type FeatureVector = ArrayVec<usize, 32>;
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply src/quantized_eg_policy.rs:33]: the 32 in ArrayVec<usize, 32> is the maximum number of pieces on a chess board, which is correct and self-contained. POLICY_NUMBER_FEATURES = 768 is the weight matrix dimension (one vector per possible
  feature index), not the maximum number of active features per position — they're unrelated quantities and coupling them would be misleading. There's nothing to guard against since ArrayVec already panics on
  overflow if the bound were ever wrong. Given this is also on the per-position inference hot path, any added indirection isn't worth it.

File: crates/princhess-train/src/bin/train-eg-policy.rs [hunk 7]
```diff
@@ -0,0 +1,1516 @@
...
+
+const LR: f32 = 1e-3;
+
+const SOFT_TARGET_WEIGHT: f32 = 0.1;
+const SOFT_TARGET_TEMPERATURE: f32 = 4.0;
+
+const EPSILON: f32 = 1e-9;
+const SAVE_EVERY_N_SUPER_BATCHES: usize = 10;
+
+const _BUFFER_SIZE_CHECK: () = assert!(TrainingPosition::BUFFER_SIZE.is_multiple_of(BATCH_SIZE));
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-eg-policy.rs]: Fixed in [29bbb2a](https://github.com/princesslana/princhess/pull/480/commits/29bbb2a2474fe4f255d115358c556de174b7c9b2)

File: crates/princhess-train/src/bin/train-eg-policy.rs [hunk 8]
```diff
@@ -0,0 +1,1516 @@
...
+        ($v:expr) => {
+            *lazy.entry($v.0).or_default() += 1;
+            clean_lo!();
+            match lo.peek() {
+                Some(&top) if $v > top => hi_eff -= 1,
+                _ => lo_eff -= 1,
+            }
+            rebalance!();
+        };
+    }
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/train-eg-policy.rs:597]: Fixed in [29bbb2a](https://github.com/princesslana/princhess/pull/480/commits/29bbb2a2474fe4f255d115358c556de174b7c9b2)

File: crates/princhess-train/src/data.rs [hunk 9]
```diff
@@ -193,6 +194,56 @@ impl From<&TrainingPosition> for State {
...
+                    return TrainingPosition::read_buffer(&self.buf);
+                }
+                Err(e) if e.kind() == ErrorKind::UnexpectedEof => {
+                    self.file.seek(SeekFrom::Start(0)).unwrap();
+                    self.positions_consumed = 0;
+                }
+                Err(e) => panic!("Data read error: {e}"),
+            }
+        }
+    }
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/data.rs:256]: Fixed in [29bbb2a](https://github.com/princesslana/princhess/pull/480/commits/29bbb2a2474fe4f255d115358c556de174b7c9b2)

File: crates/princhess-train/src/neural/lr_scheduler.rs [hunk 10]
```diff
@@ -1,36 +1,55 @@
...
     fn get_lr(&self, step: u32) -> f32 {
         if step < self.warmup_steps {
             self.initial_lr * step as f32 / self.warmup_steps as f32
         } else {
             let decay_steps = self.total_steps - self.warmup_steps;
             let elapsed = step - self.warmup_steps;
-            self.initial_lr * (1.0 - elapsed as f32 / decay_steps as f32).max(0.0)
+            self.initial_lr * (1.0 - elapsed as f32 / decay_steps as f32).max(0.0).powf(self.power)
         }
     }
```
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/neural/lr_scheduler.rs:48]: Fixed in [29bbb2a](https://github.com/princesslana/princhess/pull/480/commits/29bbb2a2474fe4f255d115358c556de174b7c9b2)

File: crates/princhess-train/src/bin/data-gen.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ianagbip1oti) [reply crates/princhess-train/src/bin/data-gen.rs:1766]: Fixed in [ff81fbf](https://github.com/princesslana/princhess/pull/480/commits/ff81fbfcdb12229d6001f5547e67cec7162a7c7e)
